### PR TITLE
[#10931][feat] AutoDeploy: one-model spec dec

### LINKED ISSUE #11549: [Feature]:  AutoDeploy: one-model speculative decoding
### 🚀 The feature, motivation and pitch

run one-model speculative decoding e2e with native attention_interface updates

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and checked the [documentation](https://nvidia.github.io/TensorRT-LLM/) and [examples](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples) for answers to frequently asked questions.

### PR DESCRIPTION
**Pull request by lucaslie**
## Summary by CodeRabbit

* **New Features**
  * Added support for Eagle3 one-model speculative decoding path, enabling single-model inference optimizations with speculative decoding capabilities.
  * Introduced `speculative_model_kwargs` configuration field for passing additional parameters to speculative models.

* **Improvements**
  * Enhanced batch metadata handling and KV cache resource management for improved inference efficiency.
  * Streamlined public APIs by consolidating internal metadata structures.

* **Bug Fixes**
  * Fixed overlap scheduler behavior and token acceptance rate tracking in speculative decoding workflows.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/auto_deploy/models/custom/modeling_eagle.py [hunk 1]
```diff
@@ -575,323 +597,308 @@ def apply_draft_embedding(self, input_ids: torch.Tensor) -> torch.Tensor:
...
+            # for idx>1 --> we offset uniformly by 1
+            if draft_idx == 1:
+                c_offset = torch.ones(num_sequences, dtype=torch.int32, device=device)
+
+            # switch to generate (if not done already), store new tokens, and offset cache
+            # can be skipped for last iteration since after we return metadata will be reset
+            if draft_idx < self.max_draft_len - 1:
+                csi.info.switch_to_generate_()
+                csi.info.copy_("input_ids", draft_tokens)
+                csi.info.offset_pos_and_cache_(c_offset)
```
Comment (lucaslie) [inline tensorrt_llm/_torch/auto_deploy/models/custom/modeling_eagle.py]: TODO: I believe this offset should be applied at the BEGINNING of the for-loop.

Moreover, for draft_idx == 1:
1. prefill --> advance from 0 to seq_len
2. extend --> advance by 1
for draft_idx > 1:
1. prefill --> advance by 1
2. extend --> advance by 1

File: tests/integration/defs/accuracy/test_llm_api_autodeploy.py [hunk 2]
```diff
@@ -212,6 +213,85 @@ def test_attention_dp(self, world_size):
             task.evaluate(llm, sampling_params=sampling_params)
 
 
+class TestLlama3_1_8B_Instruct_Eagle3(LlmapiAccuracyTestHarness):
```
Comment (lucaslie) [inline tests/integration/defs/accuracy/test_llm_api_autodeploy.py]: @govind-ramnarayan, I am still not getting expected accuracy: 
```
===========================================================
= ACCURACY HYPOTHESIS TESTING
===========================================================
Alpha (Type I:  False Positive): 0.050
Beta  (Type II: False Negative): 0.200
Sigma (Standard deviation): 50.000
#Samples: 1319
Higher is better: True
Theta (Minimum detectable effect): 4.841
Reference accuracy: 74.200
Threshold: 70.997
===========================================================
Evaluated accuracy: 64.746
===========================================================
Spec dec acceptance rate: 7.80% (24802/317814 tokens)
```

vs what you reported to me on slack a while ago

```
===========================================================
= ACCURACY HYPOTHESIS TESTING
===========================================================
Alpha (Type I:  False Positive): 0.050
Beta  (Type II: False Negative): 0.200
Sigma (Standard deviation): 50.000
#Samples: 1319
Higher is better: True
Theta (Minimum detectable effect): 4.841
Reference accuracy: 74.200
Threshold: 70.997
===========================================================
Evaluated accuracy: 73.465
===========================================================
Spec dec acceptance rate: 31.00% (63494/204843 tokens)
```

do you think do you have some bandwidth looking into this? Maybe you can do a side-by-side comparison with Cursor/Claude about metadata/accepted token/metadata updates/etc... and see if there are any discrepancies?

**Review by govind-ramnarayan — APPROVED** [review]

**Review by govind-ramnarayan — APPROVED** [review]

**Comment by lucaslie** [comment]
/bot run --extra-stage "DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-1" --disable-fail-fast

File: tests/integration/defs/accuracy/test_llm_api_autodeploy.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lucaslie) [reply tests/integration/defs/accuracy/test_llm_api_autodeploy.py]: Hi, the test can run on pre-hopper and I don't want to skip it so we can run it locally on pre-hopper GPUs. I only register it on Hopper to save CI resources. Is that fine to leave as is ?

File: tests/integration/test_lists/test-db/l0_h100.yml [hunk 3]
```diff
@@ -442,6 +442,7 @@ l0_h100:
   - unittest/_torch/auto_deploy/unit/singlegpu
   - accuracy/test_llm_api_autodeploy.py::TestLlama3_1_8B::test_auto_dtype[trtllm-False-1]
   - accuracy/test_llm_api_autodeploy.py::TestLlama3_1_8B::test_auto_dtype[trtllm-True-1]
+  - accuracy/test_llm_api_autodeploy.py::TestLlama3_1_8B_Instruct_Eagle3::test_eagle3_one_model
```
[the remark this answers is not in the document]
Comment (lucaslie) [reply tests/integration/test_lists/test-db/l0_h100.yml]: will do 👍

**Review by xinhe-nv — APPROVED** [review]
